### PR TITLE
Chapter5

### PR DESCRIPTION
**Pull request by NguyenThu2808**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -5,6 +5,7 @@ ruby "2.4.1"
 
 # Bundle edge Rails instead: gem "rails", github: "rails/rails"
```
Comment (huuphuc93) [inline Gemfile]: remove all comment in this file

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -1,2 +1,11 @@
 module ApplicationHelper
+  # Returns the full title on a per-page basis.
```
Comment (huuphuc93) [inline app/helpers/application_helper.rb]: remove this line

File: app/views/layouts/_footer.html.erb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<footer class="footer">
+  <small>
+    The <a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
```
Comment (huuphuc93) [inline app/views/layouts/_footer.html.erb]: Use gem I18n, move text in file en.yml and translation to vienamese in vi.yml,  Use gem config move "http://www.railstutorial.org/" to file settings

File: app/views/layouts/_rails_default.html.erb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<%= csrf_meta_tags %>
+<%= csp_meta_tag %>
+<%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
```
Comment (huuphuc93) [inline app/views/layouts/_rails_default.html.erb]: use " instead of '

File: app/views/static_pages/about.html.erb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  <body>
+    <h1>About</h1>
+    <p>
+      The <a href="http://www.railstutorial.org/"><em>Ruby on Rails
+      Tutorial</em></a> is a
+      <a href="http://www.railstutorial.org/book">book</a> and
+      <a href="http://screencasts.railstutorial.org/">screencast series</a>
+      to teach web development with
+      <a href="http://rubyonrails.org/">Ruby on Rails</a>.
+      This is the sample application for the tutorial.
```
Comment (huuphuc93) [inline app/views/static_pages/about.html.erb]: Use I18n

File: config/routes.rb [hunk 6]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
+  root "static_pages#home"
+  get "/help", to: "static_pages#help"
+  get "/about", to: "static_pages#about"
+  get "/contact", to: "static_pages#contact"
+  get "/signup", to: "users#new"
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
```
Comment (huuphuc93) [inline config/routes.rb]: remove this comment

File: app/views/users/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+<% provide :title, "Sign up" %>
+<h1>Sign up</h1>
+<p>This will be a signup page for new users.</p>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb]: I18n

**Comment by NguyenThu2808** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+p {
+  font-size: 1.1em;
+  line-height: 1.7em;
+}
+/* header */
+#logo {
+  float: left;
+  margin-right: 10px;
+  font-size: 1.7em;
+  color: #fff;
```
Comment (huuphuc93) [inline app/assets/stylesheets/custom.scss]: use color variable

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -1,2 +1,10 @@
 module ApplicationHelper
+  def full_title page_title = ""
+    base_title = "Ruby on Rails Tutorial Sample App"
```
Comment (huuphuc93) [inline app/helpers/application_helper.rb]: I18n

File: app/views/layouts/_footer.html.erb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+<footer class="footer">
+  <small>
+    <%= t ".the" %> <a href="http://www.railstutorial.org/"><%= t ".railstutorial" %></a>
```
Comment (huuphuc93) [inline app/views/layouts/_footer.html.erb]: move "http://www.railstutorial.org/" to file Settings, use gem config

File: app/views/static_pages/about.html.erb [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<% provide :title, "About" %>
+<!DOCTYPE html>
+<html>
+  <head>
+    <title><%= yield(:title) %> | <%= t ".railstutorialapp" %></title>
+  </head>
+  <body>
+    <h1><%= t ".title" %></h1>
+    <p>
+      <%= t ".the" %> <a href="http://www.railstutorial.org/"><em><%= t ".railstutorial" %></em></a> <%= t ".isa" %>
```
Comment (huuphuc93) [inline app/views/static_pages/about.html.erb]: move "http://www.railstutorial.org/" to Settings

File: app/views/static_pages/contact.html.erb [hunk 12]
```diff
@@ -0,0 +1,6 @@
+<% provide :title, "Contact" %>
```
Comment (huuphuc93) [inline app/views/static_pages/contact.html.erb]: I18n

File: app/views/static_pages/help.html.erb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<% provide :title, "Help" %>
```
Comment (huuphuc93) [inline app/views/static_pages/help.html.erb]: I18n

File: app/views/static_pages/home.html.erb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+<% provide :title, "Home" %>
```
Comment (huuphuc93) [inline app/views/static_pages/home.html.erb]: I18n

File: app/views/users/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,3 @@
+<% provide :title, "Sign up" %>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb]: I18n

File: config/locales/en.yml [hunk 16]
```diff
@@ -1,33 +1,49 @@
-# Files in the config/locales directory are used for internationalization
```
Comment (huuphuc93) [inline config/locales/en.yml:1]: add file vi.yml to use vietnamese in I18n

**Comment by huuphuc93** [comment]
commented

**Comment by NguyenThu2808** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 17]
```diff
@@ -0,0 +1,80 @@
+@import "bootstrap-sprockets";
+@import "bootstrap";
+@import "color.scss";
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Use single quote in CSS

File: app/assets/stylesheets/custom.scss [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+/* universal */
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Add a line break before this comment

File: app/assets/stylesheets/custom.scss [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+body {
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Add a line break after the comment

File: app/assets/stylesheets/custom.scss [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+textarea {
+  resize: vertical;
+}
+.center {
+  text-align: center;
+  h1 {
+    margin-bottom: 10px;
+  }
+}
+/* typography */
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Same here

File: app/assets/stylesheets/custom.scss [hunk 21]
```diff
@@ -0,0 +1,80 @@
+@import "bootstrap-sprockets";
+@import "bootstrap";
+@import "color.scss";
+/* universal */
+body {
+  padding-top: 60px;
+}
+section {
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Add a line break before starting a new block

File: app/assets/stylesheets/custom.scss [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+  }
+  ul {
+    float: right;
+    list-style: none;
+    li {
+      float: left;
+      margin-left: 15px;
+    }
+  }
+}
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss:94]: Please do the following for this file
1. Add a line break before and after the comment
2. Add a line break before starting a block

File: app/controllers/static_pages_controller.rb [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+  def home
+  end
+
+  def help
+  end
+
+  def about
+  	
+  end
+  def contact
```
Comment (mdmahmudurrahman) [inline app/controllers/static_pages_controller.rb]: Add a line break before this

File: app/controllers/static_pages_controller.rb [hunk 24]
```diff
@@ -0,0 +1,14 @@
+class StaticPagesController < ApplicationController
+  def home
+  end
+
+  def help
+  end
+
+  def about
+
```
Comment (mdmahmudurrahman) [inline app/controllers/static_pages_controller.rb]: remove this empty line

File: app/controllers/static_pages_controller.rb [hunk 25]
```diff
@@ -0,0 +1,14 @@
+class StaticPagesController < ApplicationController
+  def home
+  end
```
Comment (mdmahmudurrahman) [inline app/controllers/static_pages_controller.rb]: If an action don't have a body, declare it like as follows
```ruby
def home; end 
```

File: app/controllers/users_controller.rb [hunk 26]
```diff
@@ -0,0 +1,4 @@
+class UsersController < ApplicationController
+  def new
+  end
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb:3]: `def new; end`

File: app/models/user.rb [hunk 27]
```diff
@@ -0,0 +1,4 @@
+class User < ApplicationRecord
+	validates :name, presence: true, length: { maximum: 50 }
+	validates :email, presence: true, length: { maximum: 235 }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: 1. 2 space indentation
2. I found that, you have installed "config" gem. Can you use it to manage your constant.

File: app/views/layouts/_header.html.erb [hunk 28]
```diff
@@ -0,0 +1,12 @@
+<header class="navbar navbar-fixed-top navbar-inverse">
+  <div class="container">
+    <%= link_to t(".button_link_home"), "#", id: "logo" %>
+    <nav>
+      <ul class="nav navbar-nav navbar-right">
+        <li><%= link_to t(".button_home"), root_path%></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: Add a white space before closing the erb tag
```ruby
<%= link_to t(".button_home"), root_path %>
```

File: app/helpers/application_helper.rb (hunk 9 above)
Comment (mdmahmudurrahman) [inline app/helpers/application_helper.rb]: Implement I18n. 
Reference: [https://guides.rubyonrails.org/i18n.html](https://guides.rubyonrails.org/i18n.html)

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by NguyenThu2808** [comment]
ready

File: app/helpers/application_helper.rb [hunk 29]
```diff
@@ -1,2 +1,10 @@
 module ApplicationHelper
+  def full_title page_title = ""
+    base_title = Settings.basetitle
```
Comment (huuphuc93) [inline app/helpers/application_helper.rb]: Settings.basetitle use gem I18n not gem config

File: app/views/static_pages/about.html.erb [hunk 30]
```diff
@@ -0,0 +1,16 @@
...
+<!DOCTYPE html>
+<html>
+  <head>
+    <title><%= yield(:title) %> | <%= t ".railstutorialapp" %></title>
+  </head>
+  <body>
+    <h1><%= t ".title" %></h1>
+    <p>
+      <%= t ".the" %> <a href=<%= Settings.railstutorial %> ><em><%= t ".railstutorial" %></em></a> <%= t ".isa" %>
+      <a href=<%= Settings.bookrails %> ><%= t ".book" %></a> <%= t ".and" %>
```
Comment (huuphuc93) [inline app/views/static_pages/about.html.erb]: use link_to instead of a tag

**Comment by huuphuc93** [comment]
commented

**Comment by NguyenThu2808** [comment]
ready

**Comment by huuphuc93** [comment]
commented
add content in file vi.yml like file en.yml

**Comment by NguyenThu2808** [comment]
ready

**Comment by hungnh103** [comment]
commented
It seems be okay. You can merge now ;)

**Pull request merged**